File: gtm/polar.py
"""
Recursive Polar (Hyperspherical) Decomposition for GTM.

Converts a D-dim Cartesian vector into (radius, D-1 angles) representation,
enabling independent scalar quantization of each hyperspherical coordinate.
"""
import torch
import math
from typing import Tuple


def polar_transform(v: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Convert Cartesian vector(s) to hyperspherical coordinates.

    Args:
        v: (..., D)
    Returns:
        r:      (...,)     - L2 norm (radius)
        thetas: (..., D-1) - angles in [0, pi] except last in [0, 2*pi)
    """
    eps = 1e-8
    D = v.shape[-1]
    r = torch.norm(v, dim=-1).clamp(min=eps)          # (...,)
    v_norm = v / r.unsqueeze(-1)                       # (..., D)  unit vector

    thetas = []
    # recursive: theta_i = arccos(x_i / sqrt(x_i^2 + ... + x_D^2))
    running_sq = (v_norm ** 2).flip(-1).cumsum(-1).flip(-1)  # (..., D)

    for i in range(D - 1):
        numer = v_norm[..., i]
        denom = running_sq[..., i].clamp(min=eps).sqrt()
        cos_theta = (numer / denom).clamp(-1.0, 1.0)
        theta = torch.acos(cos_theta)                  # (...,) in [0, pi]
        # Last angle: extend to [0, 2*pi) using sign of last component
        if i == D - 2:
            sign = torch.sign(v_norm[..., -1])
            sign[sign == 0] = 1.0
            theta = torch.where(sign < 0, 2 * math.pi - theta, theta)
        thetas.append(theta)

    thetas = torch.stack(thetas, dim=-1)               # (..., D-1)
    return r, thetas


def polar_inverse(r: torch.Tensor, thetas: torch.Tensor) -> torch.Tensor:
    """
    Convert hyperspherical coordinates back to Cartesian.

    Args:
        r:      (...,)
        thetas: (..., D-1)
    Returns:
        v: (..., D)
    """
    D = thetas.shape[-1] + 1
    sin_prods = torch.ones(*thetas.shape[:-1], device=thetas.device, dtype=thetas.dtype)  # (...,)
    coords = []
    for i in range(D - 1):
        theta_i = thetas[..., i]
        coords.append(sin_prods * torch.cos(theta_i))
        sin_prods = sin_prods * torch.sin(theta_i)
    coords.append(sin_prods)                           # last: product of all sines
    v_unit = torch.stack(coords, dim=-1)               # (..., D)
    return v_unit * r.unsqueeze(-1)

File: omniaquant-apex/src/gtm/polar.rs
//! Recursive Polar (Hyperspherical) Decomposition for GTM.
//!
//! Converts a D-dim Cartesian vector into (radius, D-1 angles) representation,
//! enabling independent scalar quantization of each hyperspherical coordinate.

use ndarray::{Array1, ArrayView1};
use std::f64::consts::PI;

const EPS: f64 = 1e-8;

/// Convert Cartesian vector to hyperspherical coordinates.
///
/// Returns:
/// - r: L2 norm (radius)
/// - thetas: D-1 angles in [0, pi] except last in [0, 2*pi)
pub fn polar_transform(v: ArrayView1<f64>) -> (f64, Array1<f64>) {
    let d = v.len();
    let r = v.dot(&v).sqrt().max(EPS);

    // Normalized unit vector
    let v_norm: Array1<f64> = v.mapv(|x| x / r);

    // Compute running squared sums from the end
    let mut running_sq = vec![0.0f64; d];
    let mut cumsum = 0.0f64;
    for i in (0..d).rev() {
        cumsum += v_norm[i] * v_norm[i];
        running_sq[i] = cumsum;
    }

    let mut thetas = Vec::with_capacity(d - 1);

    for i in 0..(d - 1) {
        let numer = v_norm[i];
        let denom = running_sq[i].max(EPS).sqrt();
        let cos_theta = (numer / denom).clamp(-1.0, 1.0);
        let mut theta = cos_theta.acos();

        // Last angle: extend to [0, 2*pi) using sign of last component
        if i == d - 2 {
            let sign = if v_norm[d - 1] < 0.0 { -1.0 } else { 1.0 };
            if sign < 0.0 {
                theta = 2.0 * PI - theta;
            }
        }

        thetas.push(theta);
    }

    (r, Array1::from_vec(thetas))
}

/// Convert hyperspherical coordinates back to Cartesian.
pub fn polar_inverse(r: f64, thetas: &Array1<f64>) -> Array1<f64> {
    let d = thetas.len() + 1;
    let mut coords = Vec::with_capacity(d);
    let mut sin_prod = 1.0f64;

    for i in 0..(d - 1) {
        let theta_i = thetas[i];
        coords.push(sin_prod * theta_i.cos());
        sin_prod *= theta_i.sin();
    }

    // Last coordinate: product of all sines
    coords.push(sin_prod);

    // Scale by radius
    Array1::from_vec(coords).mapv(|x| x * r)
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_abs_diff_eq;
    use ndarray::Array1;

    #[test]
    fn test_polar_norm_preservation() {
        let v: Array1<f64> = Array1::from_vec(vec![3.0, 4.0, 0.0, 12.0]);
        let (r, _) = polar_transform(v.view());
        assert_abs_diff_eq!(r, 13.0, epsilon = 1e-10);
    }

    #[test]
    fn test_polar_round_trip() {
        let v: Array1<f64> = Array1::from_vec(vec![1.0, -2.0, 3.0, -4.0, 5.0]);
        let (r, thetas) = polar_transform(v.view());
        let v_rec = polar_inverse(r, &thetas);

        for i in 0..v.len() {
            assert_abs_diff_eq!(v[i], v_rec[i], epsilon = 1e-10);
        }
    }

    #[test]
    fn test_zero_vector() {
        let v = Array1::<f64>::zeros(16);
        let (r, _) = polar_transform(v.view());
        assert!(r < 1e-6);
    }
}
